File: solutions/day04/solution.py
from md5 import md5

key = 'ckczppom'
def winner(d, num_zeros):
    return md5(key + str(d)).hexdigest().startswith('0' * num_zeros)

for part, num_zeros in (('part 1', 5), ('part 2', 6)):
    x = 0
    while True:
        if winner(x, num_zeros):
            print part, x
            break
        x += 1

File: solutions/day04/solution.clj
(def key "ckczppom")

(defn md5
  ; http://rosettacode.org/wiki/MD5#Clojure
  [input]
  (apply str
    (map (partial format "%02x")
      (.digest (doto (java.security.MessageDigest/getInstance "MD5")
                     .reset
                     (.update (.getBytes input)))))))

(defn hashit
  [d]
  (md5 (format "%s%s" key (str d))))

(defn winner?
  [d num-zeros]
  (every? #{\0} (take num-zeros (hashit d))))

(defn find-first-with-n-zeros 
  [num-zeros]
  (loop [i 0]
    (if (winner? i num-zeros)
      i
      (recur (inc i)))))

(println "part 1:" (find-first-with-n-zeros 5))
(println "part 2:" (find-first-with-n-zeros 6))
